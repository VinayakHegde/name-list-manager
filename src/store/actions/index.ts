import { action } from "typesafe-actions";

export enum types {
  ADD = "ADD",
  REMOVE = "REMOVE",
  PICK = "PICK",
}

export const add = (item: string) => action(types.ADD, item);
export const remove = (index: number) => action(types.REMOVE, index);
export const pick = () => action(types.PICK);
