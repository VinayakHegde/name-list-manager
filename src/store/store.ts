import { createStore } from "redux";
import rootReducer from "./reducers";

const store = createStore(rootReducer);

(window as any).store = store;

export default store;
