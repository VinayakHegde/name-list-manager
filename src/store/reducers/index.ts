import { combineReducers } from "redux";
import * as AppTypes from "AppTypes";
import { isNull } from "util";
import { types } from "../actions";
import pick from "../../utils/pick";
import { IAppState } from "../../utils/interfaces";

export const initialState: IAppState = {
  list: [],
  random: null,
};

export const reducer = (
  state: IAppState = initialState,
  action: AppTypes.AppAction
) => {
  switch (action.type) {
    case types.ADD: {
      let random = state.random;
      if (!isNull(random)) {
        random += 1;
      }
      return {
        ...state,
        list: [action.payload, ...state.list],
        random,
      };
    }
    case types.REMOVE: {
      const prevList = [...state.list];
      prevList.splice(action.payload, 1);
      const list = prevList;
      let random = state.random;

      if (!isNull(random)) {
        if (random === action.payload) {
          random = null;
        } else if (random > action.payload) {
          random -= 1;
        }
      }

      return {
        ...state,
        list,
        random,
      };
    }
    case types.PICK: {
      const random = pick(state);
      return {
        ...state,
        random,
      };
    }
    default:
      return { ...state };
  }
};

export default combineReducers({
  state: reducer,
});
