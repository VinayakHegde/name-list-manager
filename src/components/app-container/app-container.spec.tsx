import * as React from "react";
import { Provider } from "react-redux";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import store from "../../store/store";

import AppContainer from ".";

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const withProvider = () => {
  return (
    <Provider store={store}>
      <AppContainer />
    </Provider>
  );
};

describe("<AppContainer />", () => {
  it("should match the snapshot", async () => {
    act(() => {
      render(withProvider(), container);
    });
    expect(container).toMatchSnapshot();
  });
  it("should render form element", async () => {
    act(() => {
      render(withProvider(), container);
    });
    expect(container.getElementsByClassName("addItem")).toHaveLength(1);
  });
});
