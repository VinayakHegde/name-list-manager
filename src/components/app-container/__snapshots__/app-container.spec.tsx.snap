// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppContainer /> should match the snapshot 1`] = `
<div>
  <form
    class="addItem"
    data-testid="addItem"
  >
    <input
      placeholder="Enter a name"
      value=""
    />
    <button
      class="primary"
      disabled=""
      type="submit"
    >
      Add
    </button>
    <div
      class="addItem__notice"
    >
      Number of items in the list: 0
    </div>
  </form>
  <section
    class="itemList"
  />
</div>
`;
